Snackbar: First steps at potential implementation

a8ck3n opened the pull request:
### Details
Part of #938, **In-App Notifications: Snackbar**. This is the initial plumbing demonstrating one possible way to present Snackbars. Covers the view lifecycle, display, and clean up. Basic support for dynamic text.

Still needs:

- [x] 1. Proper theming for light & dark mode.
- [x] 2. Re-sizing the view based on message length. The original Issue shows the containing view only taking up the width necessary to show the message + the action button's title. Need some clarity on clipping vs wrapping longer messages.
- [x] 3. Additional initializer to wire up the action button. Currently the button just logs when clicked. Additional initializer would include a closure and execute that when clicked.
- [x] 4. Prevents additional notices if one is currently displayed.
- [x] 5. Needs to set duration based on presence of action button. Currently just puts up the notice for 2 seconds regardless of type.
- [ ] 6. Final colours for Dark & Light mode.
- [ ] 7. Haptic feedback.
- [ ] 8. Rotation support (appears to work but not throughly tested yet).
- [ ] 9. Remove debug logs showing presentation flow.

Not implemented but might be nice to have:

- [ ] Queuing mechanism if we wanted to chain notices.

### Test
1. Build and run.
2. With a note visible, click the Note Info button to see a simple message. Click the Note Options button to see a message with an action button.
3. See the Snackbar animate in then fade out.

### Release
**Do not merge!** This is a draft of work in progress.

### Comments

The basic architecture is built around the `Snackbar` and `SnackbarPresenter` classes with `SnackbarViewController` for XIB-based loading of the view. Only the `Snackbar` class is needed to present a notification.

```
let snackbar = Snackbar(message: "Some message")
snackbar.show()
```

For a Snackbar with action button:
```
let snackbar = Snackbar(message: "Snackbar with action!", actionTitle: "Boom!") {
    print("Callback code executed!")
}

```

**Light Mode example:**

<img src="https://user-images.githubusercontent.com/40267301/99146734-b7d41e00-26b5-11eb-96d8-41dbfa9c4027.gif" width="300" />

**Dark Mode with larger text:**

<img src="https://user-images.githubusercontent.com/40267301/99146778-fcf85000-26b5-11eb-9f07-8ac80d6ce91e.gif" width="300" />

a8ck3n committed:
Initial steps to Snackbar.

Introduces the Snackbar class and a private singleton to manage the lifecycle (since the views will not be part of the standard view controller hierarchy).

No UI yet. Just logging the key lifecycle events. Triggered by the Note Info button.

a8ck3n committed:
Adds a simple test view to the mix.

Shows a test view when showing and removes it when hiding.

a8ck3n committed:
Animate the view in and then fade it out.

a8ck3n committed:
Pull the text view from a XIB.

Adds a simple view controller and allows for setting up layout + constraints visually.

a8ck3n committed:
Update XIB with StackView and subviews.

Now includes the label and action button with basic layout constraints. Still needs sizing of the snackbar width if we want to shrink it to match the label + button title width.

jleandroperez commented on Simplenote/Snackbar.swift:
```diff
@@ -0,0 +1,128 @@
+				}
+			}
+		}
+	}
```
Few notes!

1. Please review encapsulation: It'd be neat if properties not meant for external usage were protected
2. Would it be possible to convert to **Autolayout**?
3. Please review Tabs / Spaces (the project itself uses Spaces rather than Tabs!)
4. Please use constants when appropriate (**UIKitConstants** may be of help!)
5. Rather than animating `asyncAfter` please consider using the `UIView.animate`'s **delay** parameter
6. What would happen if the Window rotates?
7. Please bear in mind that the action itself is optional

a8ck3n replied:
Thank you for these! I was able to address some of them in my changes today with the exceptions being:

- 2. **Move to Auto Layout** — I haven't had a chance to work on this yet. I thought it would make sense to have the view properly configured (action vs no action) before digging into this. One question, do you recommend moving away from UIStackView as part of tackling this?
- 5. **Try without `asyncAfter`** — I tested this but using `UIView.animate`'s **delay** parameter caused the button to stop accepting input. Additionally, after wiring up the button, the code path changed a little to allow short-circuiting the timeout. If I've overlooked or misunderstood something here, please let me know and I'll have another look.
- 6. **Rotation** — Currently the Snackbar doesn't respond to rotation at all. I'm thinking this would be handled by adding some Auto Layout constraints to the view once the zoom-in animation completes but I haven't tried this yet.

I've updated the issue details to show the new API (new initializer) and added an updated GIF to show both styles in action. This still needs Auto Layout, theming for Dark/Light modes, and possibly a queuing mechanism if we want to allow queuing up of Snackbar notifications.

jleandroperez commented on Simplenote/Snackbar.swift:
```diff
@@ -0,0 +1,128 @@
+	
+	var message: String
+	
+	static var presenter = SnackbarPresenter()
```
Please review this singleton (a better spot would be, perhaps, the SnackbarPresenter class itself)

a8ck3n replied:
Not sure what I was thinking here. I've moved it to the appropriate place now!

jleandroperez reviewed (changes requested):
Hi there @a8ck3n,

Thanks for the PR! Sending your way few notes. Public API looks good BTW!

a8ck3n committed:
Clean up SnackbarPresenter singleton.

Includes a private init() to prevent instances other than the shared instance.

a8ck3n committed:
Testing fade code paths.

Calling UIView.animate() directly prevents the UIButton from receiving inputs while putting it behind an asyncAfter allows the button to work correctly.

- removeView() — No asyncAfter call
- removeViewAsync() — With asyncAfter call first.

a8ck3n committed:
Hide internal properties from public API.

a8ck3n committed:
Convert tabs to spaces.

…for files modified/added in this branch.

a8ck3n committed:
Switch to UIKitConstants for calls to UIView.animate().

a8ck3n committed:
Take Snackbar message from class.

Instead of showing the default value from the XIB.

a8ck3n committed:
Disable pointer interation in XIB.

I mistakingly enabled this thining it enabled the button. Some quick testing shows this is not necessary.

a8ck3n committed:
Properly dismiss view in response to button tap.

The SnackbarPresenter acts as the delegate to the SnackbarViewController to handle taps. A tap short-circuts the timeout and fades the view immediately.

a8ck3n committed:
Removes header template cruft.

…to match newer Swift files in project.

a8ck3n committed:
Add support for setting up Snackbar action.

The button is displayed regardless but the Snackbar now has an initializer with completion and the code is called when the button is tapped.

a8ck3n committed:
Configure Snackbar based on initializer.

Simple initializer for a simple Snackbar. If the user provides a title & callback, they get an action button.

a8ck3n committed:
Set presentation duration based on presence of action.

a8ck3n committed:
Convert UIView.animate to use constraints.

Includes the following changes:

- View controller now loads two versions of the view and provides an API for configuration.
- Views use Auto Layout. No longer using a UIStackView.
- Remove code paths for simple test views.
- Position and animate the Snackbar using Auto Layout and constraints.
- Supports basic theming for Dark/Light mode (but not using the final colors yet).
- Supports Dynamic Text.

jleandroperez commented:
Closing this one for now. Thank you Kevin!
